lib/index.js

```diff
@@ -72,388 +35,341 @@ const defaultRules = [
-      super.onReadIgnoreFile(rootBuiltinRules, rules, _ => _)
-    } else {
-      this.bundled = []
-      this.bundledScopes = []
-      this.packageJsonCache = this.parent.packageJsonCache
+    // we path.resolve() here because ignore-walk doesn't do it and we want full paths
+    options.path = resolve(options.path || process.cwd()).replace(/\\/g, '/')
+
+    if (!options.ignoreFiles) {
+      options.ignoreFiles = [
+        defaultRules,
+        'package.json',
+        '.npmignore',
+        '.gitignore',
+        strictRules,
+      ]
     }
-  }
 
-  get isProject () {
-    return !this.parent || this.parent.follow && this.isSymbolicLink
+    super(options)
+    this.isPackage = options.isPackage
+    this.seen = options.seen || new Set()
+    this.tree = options.tree // no default, we'll load the tree later if we need to
+    this.requiredFiles = options.requiredFiles || []
+
+    // go ahead and inject the default rules now
+    this.injectRules(defaultRules, defaults)
+
+    if (!this.isPackage) {
+      // if this instance is not a package, then place some strict default rules, and append
+      // known required files for this directory
+      this.injectRules(strictRules, [
+        ...strictDefaults,
+        ...this.requiredFiles.map((file) => `!${file}`),
+      ])
+    }
   }
 
-  onReaddir (entries) {
-    if (this.isProject) {
-      entries = entries.filter(e =>
-        e !== '.git' &&
-        !(e === 'node_modules' && this.bundled.length === 0)
-      )
+  // overridden method: we intercept the reading of the package.json file here so that we can
+  // process it into both the package.json file rules as well as the strictRules synthetic rule set
+  addIgnoreFile (file, callback) {
+    // if we're adding anything other than package.json, then let ignore-walk handle it
+    if (file !== 'package.json' || !this.isPackage) {
+      return super.addIgnoreFile(file, callback)
     }
 
-    // if we have a package.json, then look in it for 'files'
-    // we _only_ do this in the root project, not bundled deps
-    // or other random folders.  Bundled deps are always assumed
-    // to be in the state the user wants to include them, and
-    // a package.json somewhere else might be a template or
-    // test or something else entirely.
-    if (!this.isProject || !entries.includes('package.json')) {
-      return super.onReaddir(entries)
+    return this.processPackage(callback)
+  }
+
+  // overridden method: if we're done, but we're a package, then we also need to evaluate bundles
+  // before we actually emit our done event
+  emit (ev, data) {
+    if (ev !== 'done' || !this.isPackage) {
+      return super.emit(ev, data)
     }
 
-    // when the cache has been seeded with the root manifest,
-    // we must respect that (it may differ from the filesystem)
-    const ig = path.resolve(this.path, 'package.json')
+    this.gatherBundles().then(() => {
+      super.emit('done', this.result)
+    })
+    return true
+  }
```

Comment:
Would it be possible for `ev` to be something other than done, but not want to `gatherBundles` and then override the `ev` with `"done"`?

Reply:
nope, `done` is the only event that this module will emit that we care about or want to gather bundles for